src/Pate/Verification/Widening.hs

```diff
@@ -125,37 +126,41 @@ makeFreshAbstractDomain _bak _bundle preDom (ReturnNode fPair) = do
 --   If, for some reason, we cannot find appropraite locations to widen, we
 --   widen as much as we can, and report an error.
 widenAlongEdge ::
-  SimBundle sym arch {- ^ results of symbolic execution for this block -} ->
+  SimBundle sym arch v {- ^ results of symbolic execution for this block -} ->
   GraphNode arch {- ^ source node -} ->
-  AbstractDomain sym arch {- ^ source abstract domain -} ->
+  AbstractDomain sym arch v {- ^ source abstract domain -} ->
   PairGraph sym arch {- ^ pair graph to update -} ->
   GraphNode arch {- ^ target graph node -} ->
   EquivM sym arch (PairGraph sym arch)
-widenAlongEdge bundle from d gr to = withPredomain bundle d $ \bak -> do
+widenAlongEdge bundle from d gr to = withPredomain bundle d $ \bak -> withSym $ \sym ->
 
   case getCurrentDomain gr to of
     -- This is the first time we have discovered this location
     Nothing ->
      do traceBundle bundle ("First jump to " ++ show to)
         -- initial state of the pair graph: choose the universal domain that equates as much as possible
 
-        d' <- makeFreshAbstractDomain bak bundle (PS.specBody d) to
-
-        -- compute an initial widening, if necessary
-        md <- widenPostcondition bak bundle d d'
-        case md of
-          NoWideningRequired ->
-            return (freshDomain gr to d')
-          WideningError msg _ d'' ->
-            do let msg' = ("Error during widening: " ++ msg)
-               traceBundle bundle msg'
-               return $ recordMiscAnalysisError (freshDomain gr to d'') to (Text.pack msg')
-          Widen _ _ d'' ->
-            return (freshDomain gr to d'')
+        postSpec <- makeFreshAbstractDomain bak bundle d from to
+        (asm, d') <- liftIO $ PS.bindSpec sym (PS.bundleOutVars bundle) postSpec
+        withAssumption_ (return asm) $ do
+          md <- widenPostcondition bak bundle d d'
+          case md of
+            NoWideningRequired ->
+              return (freshDomain gr to postSpec)
+            WideningError msg _ d'' ->
+              do let msg' = ("Error during widening: " ++ msg)
+                 traceBundle bundle msg'
+                 postSpec' <- abstractOverVars bundle from to postSpec d''
+                 return $ recordMiscAnalysisError (freshDomain gr to postSpec') to (Text.pack msg')
+            Widen _ _ d'' -> do
+              postSpec' <- abstractOverVars bundle from to postSpec d''
+              return (freshDomain gr to postSpec')
 
     -- have visited this location at least once before
-    Just d' ->
-     do md <- widenPostcondition bak bundle d d'
+    Just postSpec -> do
+      (asm, d') <- liftIO $ PS.bindSpec sym (PS.bundleOutVars bundle) postSpec
```

Comment:
Lets comment what these `bindSpec` calls are doing (and, more importantly, why)

Reply:
Added some comments